File: cli/internal/sdd/artifacts.go
package sdd

import (
	"os"
	"path/filepath"
	"sort"
	"strings"
)

const specFileName = "spec.md"

// discoverArtifactPaths finds the on-disk paths for every artifact kind under a
// change root. Single-file artifacts yield a one-element slice when present;
// specs are discovered by a recursive walk for files named exactly spec.md.
func discoverArtifactPaths(changeRoot string) (ArtifactPaths, error) {
	paths := emptyArtifactPaths()
	paths.Proposal = existingFile(filepath.Join(changeRoot, "proposal.md"))
	paths.Design = existingFile(filepath.Join(changeRoot, "design.md"))
	paths.Tasks = existingFile(filepath.Join(changeRoot, "tasks.md"))
	paths.ApplyProgress = existingFile(filepath.Join(changeRoot, "apply-progress.md"))
	paths.VerifyReport = existingFile(filepath.Join(changeRoot, "verify-report.md"))

	specs, err := findSpecFiles(filepath.Join(changeRoot, "specs"))
	if err != nil {
		return ArtifactPaths{}, err
	}
	paths.Specs = specs
	return paths, nil
}

// classifyArtifacts maps each discovered artifact to its completeness state.
func classifyArtifacts(changeRoot string, paths ArtifactPaths) map[string]ArtifactState {
	return map[string]ArtifactState{
		"proposal":      fileArtifactState(paths.Proposal),
		"specs":         specsArtifactState(paths.Specs, filepath.Join(changeRoot, "specs")),
		"design":        fileArtifactState(paths.Design),
		"tasks":         fileArtifactState(paths.Tasks),
		"applyProgress": fileArtifactState(paths.ApplyProgress),
		"verifyReport":  fileArtifactState(paths.VerifyReport),
	}
}

// existingFile returns a single-element slice with path when the file exists,
// otherwise an empty slice.
func existingFile(path string) []string {
	if _, err := os.Stat(path); err == nil {
		return []string{path}
	}
	return []string{}
}

// findSpecFiles recursively collects every file named exactly spec.md under
// specsRoot, sorted. A missing specs directory yields an empty list.
func findSpecFiles(specsRoot string) ([]string, error) {
	var files []string
	err := filepath.WalkDir(specsRoot, func(path string, entry os.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if !entry.IsDir() && entry.Name() == specFileName {
			files = append(files, path)
		}
		return nil
	})
	if err != nil {
		if os.IsNotExist(err) {
			return []string{}, nil
		}
		return nil, err
	}
	sort.Strings(files)
	return files, nil
}

// fileArtifactState classifies a single-file artifact: missing when absent,
// partial when present but blank, done when it has non-whitespace content.
func fileArtifactState(paths []string) ArtifactState {
	if len(paths) == 0 {
		return ArtifactMissing
	}
	if hasContent(paths[0]) {
		return ArtifactDone
	}
	return ArtifactPartial
}

// specsArtifactState classifies the specs artifact: missing when no specs dir,
// partial when the dir is non-empty but lacks usable spec.md files, done when
// every discovered spec.md has content.
func specsArtifactState(paths []string, specsRoot string) ArtifactState {
	if len(paths) == 0 {
		if entries, err := os.ReadDir(specsRoot); err == nil && len(entries) > 0 {
			return ArtifactPartial
		}
		return ArtifactMissing
	}
	for _, path := range paths {
		if !hasContent(path) {
			return ArtifactPartial
		}
	}
	return ArtifactDone
}

// hasContent reports whether the file exists and contains non-whitespace text.
func hasContent(path string) bool {
	content, err := os.ReadFile(path)
	return err == nil && strings.TrimSpace(string(content)) != ""
}

File: cli/src/ai_harness/sdd/artifacts.py
"""Artifact discovery and completeness classification."""

from __future__ import annotations

import os

from .models import (
    ARTIFACT_DONE,
    ARTIFACT_MISSING,
    ARTIFACT_PARTIAL,
    ArtifactPaths,
)

SPEC_FILE_NAME = "spec.md"


def discover_artifact_paths(change_root: str) -> ArtifactPaths:
    """Find the on-disk paths for every artifact kind under a change root.
    Single-file artifacts yield a one-element list when present; specs are
    discovered by a recursive walk for files named exactly spec.md."""
    paths = ArtifactPaths()
    paths.proposal = _existing_file(os.path.join(change_root, "proposal.md"))
    paths.design = _existing_file(os.path.join(change_root, "design.md"))
    paths.tasks = _existing_file(os.path.join(change_root, "tasks.md"))
    paths.apply_progress = _existing_file(os.path.join(change_root, "apply-progress.md"))
    paths.verify_report = _existing_file(os.path.join(change_root, "verify-report.md"))
    paths.specs = _find_spec_files(os.path.join(change_root, "specs"))
    return paths


def classify_artifacts(change_root: str, paths: ArtifactPaths) -> dict[str, str]:
    """Map each discovered artifact to its completeness state."""
    return {
        "proposal": _file_artifact_state(paths.proposal),
        "specs": _specs_artifact_state(paths.specs, os.path.join(change_root, "specs")),
        "design": _file_artifact_state(paths.design),
        "tasks": _file_artifact_state(paths.tasks),
        "applyProgress": _file_artifact_state(paths.apply_progress),
        "verifyReport": _file_artifact_state(paths.verify_report),
    }


def _existing_file(path: str) -> list[str]:
    return [path] if os.path.exists(path) else []


def _find_spec_files(specs_root: str) -> list[str]:
    """Recursively collect every file named exactly spec.md under specs_root,
    sorted. A missing or non-directory specs root yields an empty list; any other
    read error (e.g. permission denied) propagates, matching Go's WalkDir."""
    files: list[str] = []
    for dirpath, _dirnames, filenames in os.walk(specs_root, onerror=_reraise_walk_error):
        for name in filenames:
            if name == SPEC_FILE_NAME:
                files.append(os.path.join(dirpath, name))
    files.sort()
    return files


def _reraise_walk_error(error: OSError) -> None:
    """os.walk error callback. Go's filepath.WalkDir treats a missing or
    non-directory specs root as "no specs" (empty list) but propagates other read
    errors so the CLI exits 1. Mirror that: swallow not-found/not-a-dir, re-raise
    the rest (notably PermissionError)."""
    if isinstance(error, (FileNotFoundError, NotADirectoryError)):
        return
    raise error


def _file_artifact_state(paths: list[str]) -> str:
    """Classify a single-file artifact: missing when absent, partial when present
    but blank, done when it has non-whitespace content."""
    if not paths:
        return ARTIFACT_MISSING
    return ARTIFACT_DONE if _has_content(paths[0]) else ARTIFACT_PARTIAL


def _specs_artifact_state(paths: list[str], specs_root: str) -> str:
    """Classify the specs artifact: missing when no specs dir, partial when the
    dir is non-empty but lacks usable spec.md files, done when every discovered
    spec.md has content."""
    if not paths:
        try:
            with os.scandir(specs_root) as entries:
                if any(True for _ in entries):
                    return ARTIFACT_PARTIAL
        except (FileNotFoundError, NotADirectoryError):
            return ARTIFACT_MISSING
        return ARTIFACT_MISSING
    for path in paths:
        if not _has_content(path):
            return ARTIFACT_PARTIAL
    return ARTIFACT_DONE


def _has_content(path: str) -> bool:
    """Report whether the file exists and contains non-whitespace text."""
    try:
        # errors="replace" mirrors Go's lossy string(bytes): invalid UTF-8 never
        # raises, so a non-UTF-8 artifact is still classified by its content.
        with open(path, encoding="utf-8", errors="replace") as handle:
            return handle.read().strip() != ""
    except OSError:
        return False
